### PR TITLE
Improve segment_sum stability by k-way summation.

### PR DESCRIPTION
**Pull request by zhangqiaorjc**
segment_sum has numerical stability issues when it tries to sum a large number of very small numbers

Here's a simple repro
```python
import jax
import jax.numpy as jnp
import numpy as np

def compute_sums(length, device="cpu"):
  expected_sum = 2.0
  fill_value = expected_sum / length
  tensor = jnp.ones([length], dtype=jnp.float32) * fill_value
  if device == "cpu":
    tensor = jax.device_put(tensor, device=jax.devices('cpu')[0])
  elif device == "gpu":
    tensor = jax.device_put(tensor, device=jax.devices('gpu')[0])
  elif device == "tpu":
    tensor = jax.device_put(tensor, device=jax.devices('tpu')[0])
  else:
    raise RuntimeError("unsupported device")

  segment_ids = jnp.zeros([length], dtype=jnp.int32)
  reduce_sum = jnp.sum(tensor)

  seg_sum = jax.ops.segment_sum(tensor,
                                segment_ids,
                                num_segments=1,
                                indices_are_sorted=True)[0]
  unsorted_segment_sum = jax.ops.segment_sum(tensor,
                                             segment_ids,
                                             num_segments=1,
                                             indices_are_sorted=False)[0]

  print((f"length = {length}, reduce_sum = {reduce_sum},"
         f"seg_sum = {seg_sum},"
         f"unsorted_segment_sum = {unsorted_segment_sum},"))
  return jnp.abs(
      jnp.hstack([
          expected_sum - reduce_sum, expected_sum - seg_sum,
          expected_sum - unsorted_segment_sum
      ]))


lengths = np.array([1000000, 10000000, 25000000, 50000000, 75000000, 100000000])

cpu_sums = []
for length in lengths:
  cpu_sums.append(np.asarray(compute_sums(length, "cpu")))
cpu_sums = np.stack(cpu_sums, axis=0).T
worst_case_diff = cpu_sums[:, -1]
print("diff=", worst_case_diff)

# == Before this fix ==
# length = 1000000, reduce_sum = 2.000000476837158,seg_sum = 2.018077850341797,unsorted_segment_sum = 2.018077850341797,
# length = 10000000, reduce_sum = 2.0000014305114746,seg_sum = 2.129534959793091,unsorted_segment_sum = 2.129534959793091,
# length = 25000000, reduce_sum = 2.000000238418579,seg_sum = 2.0,unsorted_segment_sum = 2.0,
# length = 50000000, reduce_sum = 2.0000007152557373,seg_sum = 1.0,unsorted_segment_sum = 1.0,
# length = 75000000, reduce_sum = 2.0,seg_sum = 0.5,unsorted_segment_sum = 0.5,
# length = 100000000, reduce_sum = 2.0000009536743164,seg_sum = 0.5,unsorted_segment_sum = 0.5,
# diff= [9.536743e-07 1.500000e+00 1.500000e+00]

# == After this fix ==
# length = 1000000, reduce_sum = 2.000000476837158,segment_sum_1 = 2.018077850341797,unsorted_segment_sum_1 = 2.018077850341797,
# length = 10000000, reduce_sum = 2.0000014305114746,segment_sum_1 = 1.9812092781066895,unsorted_segment_sum_1 = 1.9812092781066895,
# length = 25000000, reduce_sum = 2.000000238418579,segment_sum_1 = 1.9934706687927246,unsorted_segment_sum_1 = 1.9934706687927246,
# length = 50000000, reduce_sum = 2.0000007152557373,segment_sum_1 = 1.9934706687927246,unsorted_segment_sum_1 = 1.9934706687927246,
# length = 75000000, reduce_sum = 2.0,segment_sum_1 = 1.9881486892700195,unsorted_segment_sum_1 = 1.9881486892700195,
# length = 100000000, reduce_sum = 2.0000009536743164,segment_sum_1 = 1.9934706687927246,unsorted_segment_sum_1 = 1.9934706687927246,
# diff= [9.536743e-07 6.529331e-03 6.529331e-03]
```

### CODE REVIEW

**Comment by zhangqiaorjc** [comment]
I added `bucket_size` argument to allow users to set it based on their needs. The default 1024 should work for most cases.

**Comment by zhangqiaorjc** [comment]
`bucket_size` is changed to None by default, so bucketing becomes an opt-in behavior.

File: jax/ops/scatter.py [hunk 1]
```diff
@@ -334,4 +341,19 @@ def segment_sum(data, segment_ids, num_segments=None,
 
   out = jnp.zeros((num_segments,) + data.shape[1:], dtype=data.dtype)
   segment_ids = jnp.mod(segment_ids, num_segments)
-  return index_add(out, segment_ids, data, indices_are_sorted, unique_indices)
+
+  num_buckets = 1 if bucket_size is None \
+                  else -(-segment_ids.size // bucket_size)  # ceil_division
```
Comment (hawkinsp) [inline jax/ops/scatter.py:346]: I wonder if it's worth adding a helper function (possibly to some common place in JAX) called `ceil_of_ratio`.

TF and XLA have a similar utility in C++.

File: jax/ops/scatter.py [hunk 2]
```diff
@@ -334,4 +341,19 @@ def segment_sum(data, segment_ids, num_segments=None,
...
+  if num_buckets == 1:
```
Comment (hawkinsp) [inline jax/ops/scatter.py:347]: In general I am in two minds about this kind of fast path. In the absence of `jit` it certainly makes sense to have the fast path. In the presence of `jit` it just makes the code more complicated!

No action required, just something to keep in mind.

File: jax/ops/scatter.py [hunk 3]
```diff
@@ -302,8 +302,12 @@ def index_update(x, idx, y, indices_are_sorted=False, unique_indices=False):
...
       x, idx, y, lax.scatter, indices_are_sorted, unique_indices)
 
-def segment_sum(data, segment_ids, num_segments=None,
-                indices_are_sorted=False, unique_indices=False):
+def segment_sum(data,
+                segment_ids,
+                num_segments=None,
+                indices_are_sorted=False,
+                unique_indices=False,
+                bucket_size=None):
```
Comment (hawkinsp) [inline jax/ops/scatter.py:310]: Consider leaving a TODO about using a non-`None` bucket size by default?

**Review by hawkinsp — APPROVED** [review]
Nice!

File: jax/ops/scatter.py (hunk 2 above)
Comment (zhangqiaorjc) [reply jax/ops/scatter.py:348]: Make sense to me. I think a default that gives correctness at reasonable performance cost is better than `None`. But we can do that in a separate PR.

File: jax/ops/scatter.py (hunk 3 above)
Comment (zhangqiaorjc) [reply jax/ops/scatter.py]: Done.

File: jax/ops/scatter.py (hunk 1 above)
Comment (zhangqiaorjc) [reply jax/ops/scatter.py]: Done.

**Pull request merged**